KAFKA-14264; New logic to discover group coordinator

philipnee opened the pull request:
[KAFKA-14264](https://issues.apache.org/jira/browse/KAFKA-14264)
In this patch, we refactored the existing FindCoordinator mechanism. In particular, we first centralize all of the network operation (send, poll) in `NetworkClientDelegate`, then we introduced a RequestManager interface that is responsible to handle the timing of different kind of requests, based on the implementation.  In this path, we implemented a `CoordinatorRequestManager` which determines when to create an `UnsentRequest` upon polling the request manager.

philipnee committed:
FindCoordinator state machine

philipnee committed:
state transition

philipnee committed:
Added some timeout logic to ensureCoordinator ready

wip

wip

philipnee committed:
Tests

philipnee committed:
fixed broken tests

philipnee committed:
Refactor network client out of the coordinator manager

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorManager.java:
```diff
@@ -0,0 +1,245 @@
+
+public class CoordinatorManager {
+    final static int RECONNECT_BACKOFF_EXP_BASE = 2;
+    final static double RECONNECT_BACKOFF_JITTER = 0.0;
```
i think we don't need jitter for the exp. backoff here.

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorManager.java:
```diff
@@ -0,0 +1,245 @@
+            // no request has been sent
+            return Optional.of(
+                    new NetworkClientUtils.UnsentRequest(
+                            this.time.timer(requestTimeoutMs), // TODO: What is the correct timer to use here
```
Here we need a timeout to expire requests that sits in the Unsent queue (See NetworkClientUtils) Not sure if requestTimeoutMs is the right timeout to use here.

philipnee committed:
wip

philipnee committed:
Expenential backoff and testing

philipnee committed:
tests

philipnee committed:
fix

philipnee committed:
fix broken test

philipnee committed:
fixed tests and cleaned up

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java:
```diff
@@ -0,0 +1,221 @@
+
+    public void doSend(UnsentRequest r) {
+        long now = time.milliseconds();
+        Node node = r.node.orElse(client.leastLoadedNode(now));
```
`leastLoadedNode` may return null if no nodes are available. We should retry later in that case.

philipnee replied:
Think we could just fail the unsent request callback and let the manager to handle the failures.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorManager.java:
```diff
@@ -0,0 +1,245 @@
+    private final Time time;
+    private final long requestTimeoutMs;
+    private Node coordinator;
+    private final BlockingQueue<BackgroundEvent> backgroundEventQueue;
```
Instead of passing a reference to the queue, perhaps can have an `ExceptionHandler` type. This avoids the tight coupling with the background thread.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorManager.java:
```diff
@@ -0,0 +1,245 @@
+    private final Logger log;
+    private final Time time;
+    private final long requestTimeoutMs;
+    private Node coordinator;
```
nit: it is conventional to have all of the `final` fields listed together.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorManager.java:
```diff
@@ -0,0 +1,245 @@
+            return Optional.empty();
+        }
+
+        return Optional.of(
```
nit: there's a bit of duplication here and line 86. Can we make a little helper?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorManager.java:
```diff
@@ -0,0 +1,245 @@
+     * @param cause why the coordinator is marked unknown
+     * @return Optional coordinator node that can be null.
+     */
+    protected Optional<Node> markCoordinatorUnknown(String cause) {
```
What is the use case where we want the old coordinator?

philipnee replied:
I'm not sure, maybe we could just make it a void type. (the signature is from the old impl)

philipnee replied:
In the background thread - we will need to check if the coordinator and disconnect the old one.

hachikuji replied:
Is it necessary to proactively disconnect the old coordinator?

philipnee replied:
Could you elaborate a bit here? Do you mean we could rely on networkclient poll to handle the disconnect?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorManager.java:
```diff
@@ -0,0 +1,245 @@
+            this.coordinator = null;
+            lastTimeOfConnectionMs = time.milliseconds();
+        } else {
+            long durationOfOngoingDisconnect = time.milliseconds() - lastTimeOfConnectionMs;
```
We need to be a little careful with measurements like this since `milliseconds()` is not guaranteed to be monotonic. No harm here I guess if the value is less than 0, but perhaps we should use `Math.max(0, XX)` at least to make the log message less confusing.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorManager.java:
```diff
@@ -0,0 +1,245 @@
+        } else {
+            long durationOfOngoingDisconnect = time.milliseconds() - lastTimeOfConnectionMs;
+            if (durationOfOngoingDisconnect > this.rebalanceTimeoutMs)
+                log.warn("Consumer has been disconnected from the group coordinator for {}ms", durationOfOngoingDisconnect);
```
Not sure if warn is suitable for this. Maybe debug?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorManager.java:
```diff
@@ -0,0 +1,245 @@
+        return Optional.ofNullable(oldCoordinator);
+    }
+
+    private AbstractRequest.Builder<?> getFindCoordinatorRequest() {
```
nit: how about `newFindCoordinatorRequest`? Also, we may as well use the more specific type `FindCoordinatorRequest.Builder`.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java:
```diff
@@ -0,0 +1,221 @@
+            handleResponse(response, e);
+        }
+
+        public void handleResponse(ClientResponse r, Throwable t) {}
```
This could be abstract if we make the class itself abstract?

philipnee replied:
I thought about it, but I think for the case of fetch request we actually don't necessary need to `handleResponse` so I thought it would be easier to make it a base class, and only extend it when needed.

hachikuji replied:
Why would we not need it for fetch requests?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThread.java:
```diff
@@ -145,19 +162,52 @@ void runOnce() {
+    }
+
+    private boolean shouldFindCoordinator() {
+        // TODO: add conditions for coordinator discovery. Example: when there are pending commits, or we have
```
If we're not using it here, let's leave it for a future PR

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThread.java:
```diff
@@ -109,34 +129,31 @@ public void run() {
+     *  iteration.
      */
     void runOnce() {
+        // TODO: we might not need the inflightEvent here
```
Can we just get rid of it if we're not using it yet?

philipnee replied:
Can we leave it to the next "minor" PR for the clean up? (per previous discussion)

hachikuji replied:
I'd be inclined to just drop it here if we are already adding a TODO.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThread.java:
```diff
@@ -145,19 +162,52 @@ void runOnce() {
             this.inflightEvent = Optional.empty();
         }
 
+
```
nit: a bunch of unneeded newlines in here

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorManager.java:
```diff
@@ -0,0 +1,245 @@
+        private long lastSentMs;
+        private long lastReceivedMs;
+        private int numAttempts;
+        public CoordinatorRequestState() {
```
nit: we can leave out the constructor and initialize the fields directly?

philipnee replied:
will move it to the findCoordinator method.

hachikuji replied:
I just mean we can do this:
```java
        private long lastSentMs = -1;
        private long lastReceivedMs = -1;
        private int numAttempts = 0;
```

philipnee committed:
Address some of the comments.

philipnee committed:
PR comments

philipnee committed:
Adding an error event handler

philipnee committed:
initialize the coordinatorRequestState directly.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorManager.java:
```diff
@@ -0,0 +1,238 @@
+
+public class CoordinatorManager {
+    final static int RECONNECT_BACKOFF_EXP_BASE = 2;
+    final static double RECONNECT_BACKOFF_JITTER = 0.0;
```
It looks like 0.2 is a typical jitter that is used.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorManager.java:
```diff
@@ -0,0 +1,238 @@
+        this.exponentialBackoff = new ExponentialBackoff(
+                config.getLong(ConsumerConfig.RECONNECT_BACKOFF_MS_CONFIG),
+                RECONNECT_BACKOFF_EXP_BASE,
+                config.getLong(CommonClientConfigs.RECONNECT_BACKOFF_MAX_MS_CONFIG),
```
nit: we may as well use `ConsumerConfig.RECONNECT_BACKOFF_MAX_MS_CONFIG` to be consistent

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorManager.java:
```diff
@@ -0,0 +1,238 @@
+    private Optional<NetworkClientDelegate.UnsentRequest> makeFindCoordinatorRequest() {
+        return NetworkClientDelegate.UnsentRequest.makeUnsentRequest(
+                this.time.timer(requestTimeoutMs),
+                newFindCoordinatorRequest(),
```
nit: since there's just one caller of `newFindCoordinatorRequest`, maybe we can inline it here?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorManager.java:
```diff
@@ -0,0 +1,238 @@
+    private final Optional<String> groupId;
+
+    private CoordinatorRequestState coordinatorRequestState = new CoordinatorRequestState();
+    private long lastTimeOfConnectionMs = -1L; // starting logging a warning only after unable to connect for a while
```
Perhaps a more accurate name given the implementation is `timeMarkedUnknownMs` or something like that?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorManager.java:
```diff
@@ -0,0 +1,238 @@
+            this.numAttempts = 0;
+        }
+
+        public void reset() {
```
Shouldn't we reset `numAttempts` here as well?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorManager.java:
```diff
@@ -0,0 +1,238 @@
+    private final long rebalanceTimeoutMs;
+    private final Optional<String> groupId;
+
+    private CoordinatorRequestState coordinatorRequestState = new CoordinatorRequestState();
```
Can this be `final` now that we have added the `reset` method?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorManager.java:
```diff
@@ -0,0 +1,238 @@
+    private final Time time;
+    private final long requestTimeoutMs;
+    private final ErrorEventHandler errorHandler;
+    private final ExponentialBackoff exponentialBackoff;
```
Could we move this to `CoordinatorRequestState`? It looks like that's the only place it is used.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorManager.java:
```diff
@@ -0,0 +1,238 @@
+            return makeFindCoordinatorRequest();
+        }
+
+        if (coordinatorRequestState.lastReceivedMs == -1 ||
```
I think we can move this check to `CoordinatorRequestState` and give it a nice name. For example, `isRequestInflight`. 

Or perhaps we could even combine all of these checks into a single check? Maybe `canSendRequest`? Then we could structure this as the following:

```java
public Optional<NetworkClientDelegate.UnsentRequest> maybeFindCoordinator() {
  if (coordinator != null) {
    return Optional.empty();
  }

  if (coordinatorRequestState.canSendRequest()) {
    return Optional.of(makeFindCoordinatorRequest());
  } 

  return Optional.empty();
}
```

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThread.java:
```diff
@@ -145,19 +166,46 @@ void runOnce() {
 
+        // TODO: Add a condition here, like shouldFindCoordinator in the future.  Since we don't always need to find
+        //  the coordinator.
+        if (coordinatorUnknown()) {
```
nit: I'd suggest moving this check into `CoordinatorManager`. We could change the name from `tryFindCoordinator` to `maybeFindCoordinator`.

philipnee replied:
hmm, but I think we do want to check the connection against the networkClient here.  Given that we want to decouple the coordinatorManager from the networkClient.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThread.java:
```diff
@@ -145,19 +166,46 @@ void runOnce() {
         // blocking.
         if (!applicationEventQueue.isEmpty() || inflightEvent.isPresent()) {
-            networkClient.poll(time.timer(0));
+            networkClientDelegate.poll();
```
nit: why don't we just set the poll timeout to 0 in this case?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThread.java:
```diff
@@ -145,19 +166,46 @@ void runOnce() {
+        // TODO: Add a condition here, like shouldFindCoordinator in the future.  Since we don't always need to find
+        //  the coordinator.
+        if (coordinatorUnknown()) {
+            coordinatorManager.tryFindCoordinator().ifPresent(networkClientDelegate::add);
```
If `CoordinatorManager` is backing off after a failure, we want to ensure that the network thread gets woken up after the backoff has expired. Perhaps instead of returning `UnsentRequest` directly, we can create a new object:

```java
class PollResult {
  int millisUntilNextEvent;
  List<UnsentRequest> requestsToSend;
}
```
Then perhaps we could structure each manager class in a similar way. Instead of `tryFindCoordinator`, we could have a generic `poll()`. For example:
```java
class CoordinatorManager {
  PollResult poll(long currentTimeMs);
}
```

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorManager.java:
```diff
@@ -0,0 +1,238 @@
+    final static double RECONNECT_BACKOFF_JITTER = 0.0;
+
+    private final Logger log;
+    private final Time time;
```
We have to be a little careful with calls to `Time.milliseconds()` since they can be expensive when overused. Usually we would structure the poll loop to compute the current time at the start of the event loop and then pass it to each component (e.g. instead of `tryFindCoordinator()`, we would use `tryFindCoordinator(currentTimeMs`).

philipnee committed:
Addressing some PR comments

philipnee committed:
Fixed some tests

philipnee committed:
minor corrections

philipnee committed:
Remove unneeded function and extra line

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorRequestManager.java:
```diff
@@ -0,0 +1,272 @@
+        }
+
+        coordinatorRequestState.updateLastFailedAttempt(currentTimeMS);
+        if (error == Errors.GROUP_AUTHORIZATION_FAILED) {
```
Why are we expecting errors in `handleSuccessFindCoordinatorResponse`?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java:
```diff
@@ -0,0 +1,260 @@
+            if (!doSend(unsent)) {
+                log.debug("No broker available to send the request: {}", unsent);
+                unsent.callback.ifPresent(v -> v.onFailure(
+                        new IllegalThreadStateException("No node available in the kafka cluster to send the request")));
```
Use retriable exception. Maybe `NETWORK_EXCEPTION`? Or a custom exception, `NoNodeAvailableException` or sth like that.

An alternative is to add to the end of the queue? We don't need to back off. Or perhaps we can use a separate queue for requests that have no target node.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java:
```diff
@@ -0,0 +1,260 @@
+        }
+
+        trySend();
+        return this.client.poll(timer.timeoutMs(), time.milliseconds());
```
After we call `poll()`, we need to iterate through unsent requests and see if the node has been disconnected. Similar to `ConsumerNetworkClient.checkDisconnects`.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java:
```diff
@@ -0,0 +1,260 @@
+
+    public abstract static class AbstractRequestFutureCompletionHandler implements RequestCompletionHandler {
+        private final RequestFuture<ClientResponse> future;
+        private ClientResponse response;
```
Drop this field. Instead, pass to `fireCompletion` from `onComplete`.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java:
```diff
@@ -0,0 +1,260 @@
+                future.raise(e);
+            } else if (response.authenticationException() != null) {
+                future.raise(response.authenticationException());
+            } else if (response.wasDisconnected()) {
```
Check if the disconnect was to the current coordinator. If so, then `markCoordinatorUnknown()`.

Alternatively (better), we can let custom callbacks call `markCoordinatorUnknown()`. This would mean the offset commit/join group handlers need a reference to the `CoordinatorManager`. We will need this anyway because we have to check for `NOT_COORDINATOR` errors in each request type sent to the coordinator.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorRequestManager.java:
```diff
@@ -0,0 +1,272 @@
+                .setKeyType(FindCoordinatorRequest.CoordinatorType.GROUP.id())
+                .setKey(this.groupId.orElse(null));
+        return NetworkClientDelegate.UnsentRequest.makeUnsentRequest(
+                this.time.timer(requestTimeoutMs),
```
I think we can centralize request timeout logic inside `NetworkClientDelegate`?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorRequestManager.java:
```diff
@@ -0,0 +1,272 @@
+            log.info("Group coordinator {} is unavailable or invalid due to cause: {}. "
+                            + "Rediscovery will be attempted.", this.coordinator, cause);
+            this.coordinator = null;
+            timeMarkedUnknownMs = time.milliseconds();
```
Pass `currentTImeMs` as parameter to this method.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorRequestManager.java:
```diff
@@ -0,0 +1,272 @@
+    private final long requestTimeoutMs;
+    private final ErrorEventHandler errorHandler;
+    private final long rebalanceTimeoutMs;
+    private final Optional<String> groupId;
```
GroupId must not be null for `FindCoordinator` request. If groupId is null, we don't need `CoordinatorRequestManager`.

philipnee committed:
Addressed PR comments and added tests around network util

Address pr comments

1. put request timeout in networkclientDelegate
2. markCoordinatorUnknown upon failing request
3. groupId == null case handling

philipnee committed:
clean up the previous commit

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorRequestManager.java:
```diff
@@ -0,0 +1,284 @@
+     */
+    protected void markCoordinatorUnknown(final String cause, final long currentTimeMs) {
+        if (this.coordinator != null) {
+            log.info("Group coordinator {} is unavailable or invalid due to cause: {}. "
```
used to be log.info, maybe log.warn is better?

hachikuji replied:
Info seems fine. It is expected for the coordinator to change over time.

philipnee committed:
Clean up.

philipnee committed:
more clean up

Documentation and clean ups.

philipnee committed:
trigger test

trigger test

philipnee committed:
clean up

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorRequestManager.java:
```diff
@@ -0,0 +1,303 @@
+            return;
+        }
+
+        if (error == Errors.GROUP_AUTHORIZATION_FAILED) {
```
Do we need to move this check before the one above?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorRequestManager.java:
```diff
@@ -0,0 +1,303 @@
+            return;
+        }
+
+        log.debug("Group coordinator lookup failed: {}", response.errorMessage());
```
This log message seems redundant given the one above.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorRequestManager.java:
```diff
@@ -0,0 +1,303 @@
+        // handle Runtime exception
+        if (t != null) {
+            log.error("FindCoordinator request failed due to {}", t.getMessage());
+            markCoordinatorUnknown("coordinator unavailable", currentTimeMs);
```
Is this necessary?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorRequestManager.java:
```diff
@@ -0,0 +1,303 @@
+        }
+
+        log.debug("Group coordinator lookup failed: {}", response.errorMessage());
+        errorHandler.handle(error.exception());
```
Do we need to pass through retriable errors? I had thought this would only be for fatal errors.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorRequestManager.java:
```diff
@@ -0,0 +1,303 @@
+    }
+
+    public void onResponse(final FindCoordinatorResponse response, Throwable t) {
+        long currentTimeMs = time.milliseconds();
```
To be consistent, why don't we pass this through `onResponse`? We can use the `receivedTimeMs` from the `ClientResponse` object.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/FindCoordinatorResponse.java:
```diff
@@ -50,6 +52,22 @@ public FindCoordinatorResponse(FindCoordinatorResponseData data) {
         this.data = data;
     }
 
+    public List<Coordinator> getCoordinatorByKey(String key) {
```
Why don't we return `Optional<Coordinator>`? I don't think it can match more than one.

philipnee replied:
Sounds good, this is here because I think the existing code checks for multiple coordinators. I thought it would be possible to match multiple nodes.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorRequestManager.java:
```diff
@@ -0,0 +1,303 @@
+
+        FindCoordinatorResponseData.Coordinator node = coordinators.get(0);
+        if (node.errorCode() != Errors.NONE.code()) {
+            coordinatorRequestState.updateLastFailedAttempt(currentTimeMs);
```
We already do this in `handleFailedCoordinatorResponse`?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorRequestManager.java:
```diff
@@ -0,0 +1,303 @@
+
+    private void handleFailedCoordinatorResponse(
+            final FindCoordinatorResponseData.Coordinator response,
+            final long currentTimeMS) {
```
nit: `currentTimeMs`?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorRequestManager.java:
```diff
@@ -0,0 +1,303 @@
+            String msg = String.format("Group coordinator lookup failed: Response should contain exactly one " +
+                    "coordinator, it has %d", coordinators.size());
+            log.error(msg);
+            errorHandler.handle(new IllegalStateException(msg));
```
In the error case, we don't need the response, just an `Exception`. Why don't we simplify this a little:
```java
 private void handleFailedCoordinatorResponse(
            final Exception exception,
            final long currentTimeMS
)
```

Then here, we can just pass the `IllegalStateException` to `handleFailedCoordinatorResponse`.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorRequestManager.java:
```diff
@@ -0,0 +1,303 @@
+        }
+    }
+
+    private void handleSuccessFindCoordinatorResponse(
```
I think we can make these names more concise. The request type is already clear in the context, so we could call this `onSuccessfulResponse` for example.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorRequestManager.java:
```diff
@@ -0,0 +1,303 @@
+                return true;
+            }
+
+            // TODO: I think there's a case when we want to resend the FindCoordinator when we haven't received
```
What case is this?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorRequestManager.java:
```diff
@@ -0,0 +1,303 @@
+        }
+
+        private boolean requestBackoffExpired(final long currentTimeMs) {
+            return (currentTimeMs - this.lastReceivedMs) >= this.backoffMs;
```
I think this overflows if `currentTimeMs` is less than `lastReceivedMs`. Could we just write this as this:
```java
return remainingBackoffMs(currentTimeMs) > 0;
```

philipnee replied:
Sounds good, but I think currentTimeMs should always >= lastReceivedMs. Trying to think of a case there...

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThread.java:
```diff
@@ -50,57 +49,83 @@ public class DefaultBackgroundThread extends KafkaThread {
-    private final Metrics metrics;
     private final ConsumerConfig config;
+    // empty if groupId is null
+    private final Optional<CoordinatorRequestManager> coordinatorManager;
```
Instead of using `Optional`, could we use a NOOP implementation?

philipnee replied:
Let's reconsider this later? I'm not sure if that would simplify the code.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThread.java:
```diff
@@ -109,101 +134,102 @@ public void run() {
     }
 
-    private long timeToNextHeartbeatMs(final long nowMs) {
+    private long timeToNextHeartbeatMs() {
```
Let's get rid of this for now. I imagine we'll have a HeartbeatManager or something like that later.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThread.java:
```diff
@@ -109,101 +134,102 @@ public void run() {
+        Objects.requireNonNull(res);
+        if (!res.unsentRequests.isEmpty()) {
+            networkClientDelegate.addAll(res.unsentRequests);
+            return Long.MAX_VALUE;
```
Why don't we just always return `timeMsTillNextPoll`? When we start handling fetches, we may be backing off fetching from some brokers while still fetching from others.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThread.java:
```diff
@@ -109,101 +134,102 @@ public void run() {
-        if (!applicationEventQueue.isEmpty() || inflightEvent.isPresent()) {
-            networkClient.poll(time.timer(0));
+        if (!applicationEventQueue.isEmpty()) {
+            networkClientDelegate.poll(time.timer(0), false);
```
We can just set `pollWaitTimeMs` to 0.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java:
```diff
@@ -0,0 +1,291 @@
+        this.activeNodes = new HashSet<>();
+    }
+
+    public List<ClientResponse> poll(Timer timer, boolean disableWakeup) {
```
Can we get rid of `disableWakeup`? We don't really need it when the polling is done in the background. Also, why don't we just use `long timeoutMs` instead of `Timer`.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java:
```diff
@@ -0,0 +1,291 @@
+        // not available)
+        // 2. poll for the results if there's any.
+        // 3. Check connection status for each node, disconnect ones that are not reachable.
+        client.wakeup();
```
Why do we wakeup? It will ensure that `poll` returns immediately regardless of the timer.

philipnee committed:
Addressed some comments

philipnee committed:
fixed a test

philipnee committed:
remove an extra field

philipnee committed:
cleaned up

philipnee committed:
Handle Retriable error

Handle Retriable error

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorRequestManager.java:
```diff
@@ -0,0 +1,289 @@
+    public void onResponse(final FindCoordinatorResponse response, final long currentTimeMs, final Throwable t) {
+        // handle Runtime exception
+        if (t != null) {
+            log.error("FindCoordinator request failed due to {}", t.getMessage());
```
Why don't we call `handleFailedCoordinatorResponse` here? Also, we should get in the habit of including the full stack trace in error messages.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorRequestManager.java:
```diff
@@ -0,0 +1,289 @@
+        return this.coordinator;
+    }
+
+    // Visible for testing
```
The logic in this class seems general and not tied to the `FindCoordinator` API. Does it make sense to pull it out? Maybe we could call it `RequestTracker` or `RequestState` or something like that?

philipnee replied:
Thanks, that make sense.

philipnee committed:
Refactor RequestState

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/RequestState.java:
```diff
@@ -0,0 +1,92 @@
+import org.apache.kafka.clients.consumer.ConsumerConfig;
+import org.apache.kafka.common.utils.ExponentialBackoff;
+
+// Visible for testing
```
We can get rid of these comments. A general-purpose javadoc would be good though.

philipnee committed:
clean up

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorRequestManager.java:
```diff
@@ -0,0 +1,214 @@
+public class CoordinatorRequestManager implements RequestManager {
+
+    private final Logger log;
+    private final Time time;
```
Seems like we can get rid of this depenence?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorRequestManager.java:
```diff
@@ -0,0 +1,214 @@
+
+    private void onSuccessfulResponse(
+            final FindCoordinatorResponseData.Coordinator coordinator,
+            final long currentTimeMS) {
```
nit: `currentTimeMs`?

philipnee replied:
sorry, must have forgotten to change it.

philipnee committed:
pr comments

philipnee committed:
documentation

philipnee committed:
documentation

philipnee committed:
make final

philipnee committed:
Documentation

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorRequestManager.java:
```diff
@@ -0,0 +1,224 @@
+    CoordinatorRequestManager(final LogContext logContext,
+                              final ErrorEventHandler errorHandler,
+                              final String groupId,
+                              final long rebalanceTimeoutMs,
```
The dependence on the rebalance timeout is a little strange if the user is using simple partition assignment. I wonder if we could hard-code a value instead. Perhaps we can log a message every one minute (say) that the coordinator is unknown?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorRequestManager.java:
```diff
@@ -0,0 +1,224 @@
+public class CoordinatorRequestManager implements RequestManager {
+
+    private final Logger log;
+    private final ErrorEventHandler errorHandler;
```
Perhaps we could call this `nonRetriableErrorHandler` or something like that to make the usage a little clearer?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorRequestManager.java:
```diff
@@ -0,0 +1,224 @@
+
+        return new NetworkClientDelegate.PollResult(
+                coordinatorRequestState.remainingBackoffMs(currentTimeMs),
+                new ArrayList<>());
```
nit: we can use `Collections.emptyList()`?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorRequestManager.java:
```diff
@@ -0,0 +1,224 @@
+    }
+
+    /**
+     * Returns the current coordinator node. It can be {@code null}.
```
I wonder if it it would be safer to return `Optional`.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThread.java:
```diff
@@ -109,101 +132,89 @@ public void run() {
-        if (!applicationEventQueue.isEmpty() || inflightEvent.isPresent()) {
-            networkClient.poll(time.timer(0));
+        if (!applicationEventQueue.isEmpty()) {
+            networkClientDelegate.poll(0);
```
It may take a little while before we are able to send a request (e.g. the connection might have reached the max inflight requests allowed). We do not want to busy loop in this case. I'd suggest using `retryBackoffMs` in this case. We can structure the logic like this.

```java
if (!applicationEventQueue.isEmpty()) {
  pollWaitTimeMs = Math.min(retryBackoffMs, pollWaitTimeMs);
}

networkClientDelegate.poll(pollWaitTimeMs);
```

hachikuji replied:
I had the context a little wrong here. I wonder if we need this at all. Instead, we should try to drain all application events before we poll the network client. And when we post a new event, we should use `wakeup()`.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorRequestManager.java:
```diff
@@ -0,0 +1,224 @@
+
+        Optional<FindCoordinatorResponseData.Coordinator> coordinator = response.getCoordinatorByKey(this.groupId);
+        if (!coordinator.isPresent()) {
+            String msg = String.format("Coordinator not found for groupId: %s", this.groupId);
```
I think this message is a little too generic. How about something like this?
> Response did not contain expected coordinator section for groupId: %s

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorRequestManager.java:
```diff
@@ -0,0 +1,224 @@
+            final Exception exception,
+            final long currentTimeMs) {
+        coordinatorRequestState.updateLastFailedAttempt(currentTimeMs);
+        markCoordinatorUnknown("coordinator unavailable", currentTimeMs);
```
Could we improve this message? Perhaps "FindCoordinator failed with exception"?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorRequestManager.java:
```diff
@@ -0,0 +1,224 @@
+        coordinatorRequestState.updateLastFailedAttempt(currentTimeMs);
+        markCoordinatorUnknown("coordinator unavailable", currentTimeMs);
+
+        if (exception == Errors.GROUP_AUTHORIZATION_FAILED.exception()) {
```
nit: could we move the retriable case to the top here so that the fatal cases are not split?

philipnee replied:
Thanks, somehow I thought this is a retriable error.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorRequestManager.java:
```diff
@@ -0,0 +1,224 @@
+     * {@link FindCoordinatorRequestHandler}. If the response was successful, a coordinator node will be updated. If the
+     * response failed due to errors, the current coordinator will be marked unknown.
+     * @param currentTimeMs current time ins ms.
+     * @param response the response, can be null.
```
Perhaps we could clarify this? It looks like the code expects the response to be non-null if the exception is null?

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/FindCoordinatorResponse.java:
```diff
@@ -50,6 +52,22 @@ public FindCoordinatorResponse(FindCoordinatorResponseData data) {
         this.data = data;
     }
 
+    public Optional<Coordinator> getCoordinatorByKey(String key) {
```
nit: usually we drop `get` from method names

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/FindCoordinatorResponse.java:
```diff
@@ -50,6 +52,22 @@ public FindCoordinatorResponse(FindCoordinatorResponseData data) {
+    public Optional<Coordinator> getCoordinatorByKey(String key) {
+        Objects.requireNonNull(key);
+        if (this.data.coordinators().isEmpty()) {
+            // version <= 3
```
Can we write a test case which covers all versions?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEvent.java:
```diff
@@ -20,10 +20,23 @@
+    public String toString() {
+        return type + " ApplicationEvent";
+    }
+    public enum Type {
```
Perhaps we can leave this for a follow-up since we are not implementing the COMMIT type here.

philipnee replied:
this is used in the stub commit application event in the prototype.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThread.java:
```diff
@@ -50,57 +49,81 @@ public class DefaultBackgroundThread extends KafkaThread {
             this.running = true;
+            this.errorEventHandler = new ErrorEventHandler(this.backgroundEventQueue);
+            String groupId = rebalanceConfig.groupId;
+            // TODO: Maybe consider a NOOP implementation
```
Let's not leave TODOs in the code. We can file jiras instead.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThread.java:
```diff
@@ -109,101 +132,89 @@ public void run() {
+        // TODO: Add a condition here, like shouldFindCoordinator in the future.  Since we don't always need to find
+        //  the coordinator.
+        if (coordinatorManager.isPresent()) {
+            pollWaitTimeMs = Math.min(pollWaitTimeMs, handlePollResult(coordinatorManager.get().poll(currentTimeMs)));
```
nit: can we move the call to `handlePollResult` to a separate line? It makes the code easier to follow.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThread.java:
```diff
@@ -109,101 +132,89 @@ public void run() {
+
+        final long currentTimeMs = time.milliseconds();
+        // TODO: This is just a place holder value.
+        long pollWaitTimeMs = 100;
```
In `ConsumerNetworkClient`, we hard-code the following:
```java
    private static final int MAX_POLL_TIMEOUT_MS = 5000;
```
Perhaps we can do the same here and drop the TODO?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThread.java:
```diff
@@ -109,101 +132,89 @@ public void run() {
+        // TODO: This is just a place holder value.
+        long pollWaitTimeMs = 100;
+
+        // TODO: Add a condition here, like shouldFindCoordinator in the future.  Since we don't always need to find
```
Can we drop this? This is already built into the call to `CoordinatorManager.poll`.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThread.java:
```diff
@@ -109,101 +132,89 @@ public void run() {
-        // TODO: implemented when heartbeat is added to the impl
-        return 100;
+    long handlePollResult(NetworkClientDelegate.PollResult res) {
+        Objects.requireNonNull(res);
```
nit: seems like overkill. Was there any reason you wanted to be careful with the result?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java:
```diff
@@ -0,0 +1,287 @@
+    }
+
+    public static class PollResult {
+        public final long timeMsTillNextPoll;
```
nit: how about `timeUntilNextPollMs`?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java:
```diff
@@ -0,0 +1,287 @@
+    private final Time time;
+    private final Logger log;
+    private final int requestTimeoutMs;
+    private boolean wakeup = false;
```
Not sure if you missed my previous comment, but we don't need this. We want `poll` to just return after wakeup, not raise the `WakeupException`.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java:
```diff
@@ -0,0 +1,287 @@
+     * assigned node, it will find the leastLoadedOne.
+     * Here it also stores all the nodes in the {@code activeNodes}, which will then be used to check for the disconnection.
+     */
+    void trySend(final long currentTimeMs) {
```
Can this be private? Seems like we can test it indirectly through a call to `poll`?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java:
```diff
@@ -0,0 +1,287 @@
+     * @param timeoutMs
+     * @return
+     */
+    public List<ClientResponse> poll(final long timeoutMs) {
```
nit: since we are computing the current time in `DefaultBackgroundThread`, perhaps we could just pass it through to `poll`?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java:
```diff
@@ -0,0 +1,287 @@
+        }
+    }
+
+    // Visible for testing
```
Similar comment. Best not to expose the innards like this. We can test behavior through `poll`.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java:
```diff
@@ -0,0 +1,287 @@
+            }
+
+            if (!doSend(unsent, currentTimeMs, unsentAndUnreadyRequests)) {
+                log.debug("No broker available to send the request: {}", unsent);
```
Can't we retry internally until the request expires? What is the advantage to propagating this back at this point?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java:
```diff
@@ -0,0 +1,287 @@
+
+        if (!unsentAndUnreadyRequests.isEmpty()) {
+            // Handle the unready requests in the next event loop
+            unsentRequests.addAll(unsentAndUnreadyRequests);
```
Instead of rebuilding the queue after each poll(), can we use an iterator with `Iterator.remove()` to remove the request after it has been successfully sent?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java:
```diff
@@ -0,0 +1,287 @@
+            Node node = iter.next();
+            iter.remove();
+            if (client.connectionFailed(node)) {
+                client.disconnect(node.idString());
```
Why do we do this?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java:
```diff
@@ -0,0 +1,287 @@
+        return client.newClientRequest(
+                node.idString(),
+                unsent.abstractBuilder,
+                time.milliseconds(),
```
nit: we can pass through `currentTimeMs` from the poll call

philipnee replied:
We do want to timing here to be fairly accurate I think, would it be better to re-request the time in the trySend and pass it down here?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java:
```diff
@@ -0,0 +1,287 @@
+        unsentRequests.add(r);
+    }
+
+    public void ready(final Node node) {
```
nit: we should resist exposing all of these if we can. I think we want a really basic api.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java:
```diff
@@ -0,0 +1,287 @@
+    }
+
+    public void addAll(final List<UnsentRequest> unsentRequests) {
+        unsentRequests.forEach(this::add);
```
nit: `addAll`?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java:
```diff
@@ -0,0 +1,287 @@
+    public List<ClientResponse> poll(final long timeoutMs) {
+        final long currentTimeMs = time.milliseconds();
+        trySend(currentTimeMs);
+        List<ClientResponse> res = this.client.poll(timeoutMs, currentTimeMs);
```
We should take into account the unsent request queue. I think actually this is what I was thinking of in the other comment in `DefaultBackgroundThread`. If we have requests that are ready to send, but we could not send them on this poll, we want to ensure that we do not wait too long before retrying. So we want something like this:
```java
long pollTimeoutMs = timeoutMs
if (!unsentRequests.isEmpty) {
  pollTimeoutMs = Math.min(retryBackoffMs, pollTimeoutMs);
}
List<ClientResponse> res = this.client.poll(timeoutMs, currentTimeMs);
```

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThread.java:
```diff
@@ -109,101 +132,89 @@ public void run() {
-        this.inflightEvent = maybePollEvent();
-        if (this.inflightEvent.isPresent()) {
-            log.debug("processing application event: {}", this.inflightEvent);
+        Optional<ApplicationEvent> event = maybePollEvent();
```
Perhaps we can revisit this in a future patch, but I think we probably want to handle all application events if we can before we call poll.

philipnee replied:
yeah i agree. I think we would do a `drainto` and process all of the events before polling. I happened to make the change here.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThread.java:
```diff
@@ -109,101 +132,89 @@ public void run() {
     private Optional<ApplicationEvent> maybePollEvent() {
-        if (this.inflightEvent.isPresent() || this.applicationEventQueue.isEmpty()) {
-            return this.inflightEvent;
+        if (this.applicationEventQueue.isEmpty()) {
```
nit: call this `pollApplicationEvent`?

philipnee committed:
Addressed comments

addressed some comments

philipnee committed:
PR Comments

philipnee committed:
comments

philipnee committed:
checkDisconnect needs to remove disconnected requests

philipnee committed:
Add a test for FindCoordinatorResponse.coordinatorByKey

philipnee committed:
Fixed inaccurate documentation.

philipnee committed:
Clean up doSend logic

philipnee committed:
clean up the test

hachikuji reviewed (approved):
LGTM
